### PR TITLE
Merge split UPE tests with main gateway tests

### LINKED ISSUE #7953: Merge unit and E2E tests for gateways after UPE cleanup
After we streamlined the code for gateways in https://github.com/Automattic/woocommerce-payments/issues/7782, there are few tests areas that need alignment as well: unit tests and E2E tests.

### Acceptance criteria
* `test-class-upe-split-payment-gateway.php` and `test-class-upe-payment-gateway.php` classes are merged into `test-class-wc-payment-gateway-wcpay.php`
* no duplicated E2E tests (e.g. `shopper-upe-enabled-all-flows.spec.js` might have some)

### PR DESCRIPTION
**Pull request by timur27**
Closes https://github.com/Automattic/woocommerce-payments/issues/7953

#### Changes proposed in this Pull Request
This PR merges three unit test suites into one, making `test-class-wc-payment-gateway-wcpay.php` the only test suite that covers everything. While combining, several improvements were made to the tests. The most significant one is we don't mock the unit under tests now.

Additionally, this PR removes the duplicates from E2E tests. Previously, we had multiple test suites checking e.g. saving cards and each test suite was doing it under different setting e.g. UPE or legacy card. Since we streamlined them now, there's no need to have duplicated tests either.

Tests that, in my view, need further analysis to comprehend their outcomes are accompanied by my comments. These comments detail whether the test was removed, relocated to a file, or modified. If a test doesn't have a comment, this means it was simply moved to the main gateway test suite w/o any big modifications, but of course feel free to raise any questions you might have. 

#### Testing instructions

* run https://github.com/Automattic/woocommerce-payments/actions/workflows/e2e-pull-request.yml and confirm tests are passing on this branch

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7954` or branch name `merge/upe-non-upe-tests` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=merge/upe-non-upe-tests) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: fc9efa6cf040dd6f708267fba29d899489e88bc8
- Build time: 2024-01-09 13:31:50 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: tests/unit/payment-methods/test-class-upe-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-payment-gateway.php:378]: > 💡 The below applies also to `test_process_payment_returns_correct_redirect_when_using_saved_payment` in the split UPE test file.

This test was initially introduced in https://github.com/Automattic/woocommerce-payments/pull/2054/files and its main purpose was to test the `else` flow which was executed only if a token was present. There's no such logic anymore, and it wasn't there [with the UPE](https://github.com/Automattic/woocommerce-payments/blob/6.6.0/includes/payment-methods/class-upe-payment-gateway.php#L604-L606) for some time even before the refactoring (notice there's no `elseif ( $token )` check but rather the flow executes always). 

Hence, the test was removed and instead, there's a similar test `test_process_payment_returns_correct_redirect` in `test-class-wc-payment-gateway-wcpay.php` which checks the return result but doesn't focus on the saved tokens since it's irrelevant.

File: tests/unit/payment-methods/test-class-upe-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-payment-gateway.php:380]: Moved to `test-class-wc-payment-gateway-wcpay.php`

File: tests/unit/payment-methods/test-class-upe-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-payment-gateway.php:402]: Moved to `test-class-wc-payment-gateway-wcpay.php`

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:413]: Same as in `test-class-upe-payment-gateway.php`

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:591]: Moved to `test-class-wc-payment-gateway-wcpay.php`

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:374]: Moved to `tests/unit/test-class-wc-payment-gateway-wcpay.php`

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:379]: Moved to `test-class-wc-payment-gateway-wcpay.php ` as `test_should_not_use_stripe_platform_on_checkout_page_for_non_card`

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:402]: Moved to `test-class-wc-payment-gateway-wcpay.php`

```php
test_is_mandate_data_required_card_and_link()
test_is_mandate_data_required_sepa()
test_is_mandate_data_required_returns_false()
```

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:411]: Migrated to `test-class-wc-payment-gateway-wcpay.php` as `test_non_reusable_gateways_not_available_when_changing_payment_method_for_card`. The new test now checks the checkout page for changing the payment method for subscription instead of purchasing a subscription. I did so because
* setting up cart with subscription is difficult with the current design and when you're not using mocks anymore
* the new test still accomplishes the main goal (check that a non-reusable gateway is not showing up when we're working with subscriptions)

File: tests/unit/payment-methods/test-class-upe-split-payment-gateway.php
Comment (timur27) [inline tests/unit/payment-methods/test-class-upe-split-payment-gateway.php:469]: This test was removed because `wc_payment_intent_id` does not exist anymore, and was previously used only for the update payment intent flow in the UPE & split UPE implementations.

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7954` or branch name `merge/upe-non-upe-tests` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=merge/upe-non-upe-tests) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: d750ad7fbcf733f8509e581b0df190177f0f1c81
- Build time: 2024-01-19 07:30:28 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: .github/workflows/e2e-pull-request.yml
Comment (timur27) [inline .github/workflows/e2e-pull-request.yml:56]: I've merged the upe/bnpl tests into `wcpay/shopper` because I don't see the need of distinguishing between those groups anymore. One downside of this is that `wcpay - shopper` now runs a bit longer. Looking forward to hearing the opinions of others.

Same for `.github/workflows/e2e-tests-atomic.yml` changes below.

File: tests/e2e/specs/upe-split/shopper/shopper-deferred-intent-creation-upe-enabled.spec.js
Comment (timur27) [inline tests/e2e/specs/upe-split/shopper/shopper-deferred-intent-creation-upe-enabled.spec.js]: Duplicate of `shopper-myaccount-payment-methods-add-delete-success.spec.js::should add the card as a new payment method`

File: tests/e2e/specs/upe-split/shopper/shopper-upe-enabled-all-flows.spec.js
Comment (timur27) [inline tests/e2e/specs/upe-split/shopper/shopper-upe-enabled-all-flows.spec.js:1]: This test suite consist of duplicates similarly to `tests/e2e/specs/upe-split/shopper/shopper-deferred-intent-creation-upe-enabled.spec.js`.

File: tests/e2e/specs/upe-split/shopper/shopper-deferred-intent-creation-upe-enabled.spec.js
Comment (timur27) [inline tests/e2e/specs/upe-split/shopper/shopper-deferred-intent-creation-upe-enabled.spec.js:115]: Moved to `shopper-checkout-save-card-and-purchase.spec.js`.

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:1]: Removed duplicated tests and added the new `should successfully place order with Bancontact`.

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:115]: Moved to `shopper-checkout-save-card-and-purchase.spec.js.`

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:57]: This test is now a duplicate of `shopper-checkout-purchase.spec.js::using a basic card` hence is being removed.

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:66]: This test is now a duplicate of `shopper-checkout-purchase.spec.js::using a 3DS card` hence is being removed.

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:79]: This test is now a duplicate of `shopper-checkout-save-card-and-purchase.spec.js::should save the card` hence is being removed.

File: tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js
Comment (timur27) [inline tests/e2e/specs/wcpay/shopper/shopper-checkout-purchase-with-upe-methods.spec.js:109]: Duplicates of

* `shopper-checkout-save-card-and-purchase.spec.js::should process a payment with the saved card`
* `shopper-checkout-save-card-and-purchase.spec.js::should delete the card`
thus removing them.

**Comment by timur27** [comment]
`wcpay - shopper` e2e tests fail because of https://github.com/Automattic/woocommerce-payments/issues/7980. If there's a need this, one might want to push 8f0cd409cc40df3e64322a863c0161032dd38b9c to this PR and verify that all the checks are passing.

**Comment by timur27** [comment]
`wcpay - shopper` [fails on `develop`](https://github.com/Automattic/woocommerce-payments/actions/runs/7499605500/job/20416704742) with the same error.

File: .github/workflows/e2e-pull-request.yml
Comment (achyuthajoy) [reply .github/workflows/e2e-pull-request.yml:56]: @timur27 I agree that we don't need to have too many sub groups for E2E tests. Moving UPE tests to shopper makes sense as long as it doesn't add too much time addition to the test runtime.

What is the time difference you've noticed after moving the tests to shopper?

File: tests/unit/test-class-wc-payment-gateway-wcpay.php
Comment (mdmoore) [inline tests/unit/test-class-wc-payment-gateway-wcpay.php]: While I do feel like this class should be named after you @timur27 , I wonder if you meant to change this back 🤣.

File: tests/e2e/utils/payments.js
Comment (mdmoore) [inline tests/e2e/utils/payments.js]: This is really nice refactor and makes this util much more useful! The description could be updated now that it can handle multiple payment methods.

File: tests/e2e/utils/payments.js
Comment (mdmoore) [inline tests/e2e/utils/payments.js:325]: Could we update the description here as well?

**Review by mdmoore — APPROVED** [review]
This is a massive undertaking but it's really nice to see all of this consolidation! Thanks for working on it. 

As we discussed, the E2E tests are failing on `develop` so it's hard to understand if the changes here are currently passing. But I can see they were passing in [previous runs](https://github.com/Automattic/woocommerce-payments/actions/runs/7490228866), so I think it's likely that the changes here aren't contributing to the failures on the latest runs. As you mentioned, you don't intend to merge this until the E2E tests are passing on `develop`.

The unit tests are passing and the changes seem solid from what I can see(there's a lot). I left a couple really small comments/text changes. I'll go ahead and approve as those changes aren't blockers to anything. Really nice work!

File: tests/unit/test-class-wc-payment-gateway-wcpay.php
Comment (timur27) [reply tests/unit/test-class-wc-payment-gateway-wcpay.php]: Oh no 🤣 I'm so sorry, updated the class name back in 809d05dbb87875648f2df0c45a47d6ee77511c3a :sweat_smile:

File: tests/e2e/utils/payments.js
Comment (timur27) [reply tests/e2e/utils/payments.js]: Thank you! 809d05dbb87875648f2df0c45a47d6ee77511c3a

File: tests/e2e/utils/payments.js
Comment (timur27) [reply tests/e2e/utils/payments.js:325]: 👍 done in 809d05dbb87875648f2df0c45a47d6ee77511c3a, thanks!

File: .github/workflows/e2e-pull-request.yml
Comment (timur27) [reply .github/workflows/e2e-pull-request.yml:56]: @achyuthajoy thanks, I think it's not more than 110 seconds but it's hard to confirm as of today because tests are failing on `develop`. I will get back if after fixing `develop` this branch will be extended by more than ~2mins but in the meantime let's assume this is the change.

Comment (achyuthajoy) [reply .github/workflows/e2e-pull-request.yml:56]: Sound good to me.

**Comment by timur27** [comment]
Blocked by failing the e2e test on `develop.` Once those are fixed, this PR will be merged.

**Pull request merged**
